### ISSUE TITLE
Plugin for sites not using Jekyll

### ISSUE THREAD

**Issue by seshrs** [eecs485staff/primer-spec#16]
Tl;dr: I want to add a custom JavaScript "plugin" that can easily add Primer Spec to HTML websites without having to use Jekyll. Where should such a script live?

---

The _long_ version of the context:

Dr. Kapritsos expressed interest in using Primer Spec with EECS 482 specs. However, their setup involves hosting plain HTML files on a private web server accessible only to students enrolled in the course.
I wrote a quick "plugin" script that adds the scaffolding that Jekyll would have otherwise provided. To demonstrate, I uploaded the Primer Spec theme files in a [demo GitHub repo](https://github.com/seshrs/ci-playground/tree/master/docs/assets) (and also uploaded the plugin code), and then created two test pages [`plain.html`](https://seshrs.ml/beta/plain.html) and [`spec.html`](https://seshrs.ml/beta/spec.html). The two pages are identical except that the plugin is added as a `script` tag at the end of `spec.html`.

Where should this plugin live? Here are two options I thought of:

### Option 1
Add custom plugin as an integration option built into Primer Spec.

*Pros:*
  - Plugin code would live in this repo and would share code with the Jekyll theme (so they would always be in sync).
  - Our current Jekyll theme can be modified to completely use the custom plugin. This could slightly reduce complexity introduced by Jekyll. This would also cause the Jekyll layout `spec.html` to look more similar to the upstream (GitHub's original Primer theme), possibly making maintenance easier.

*Cons:*
  - Added complexity. There will be more moving pieces.
  - We would have to generate two JavaScript targets.

### Option 2
Custom plugin is maintained in a separate repository.

_Pros:_
  - Separation of concerns. This repo would be dedicated as a Jekyll Theme.
  - The plugin (which would not be used by EECS 485) would not necessarily have to be maintained by `eecs485staff`.

_Cons:_
  - Additional maintenance work might be needed to keep the external plugin up-to-date with the Jekyll theme.
  - Concerns about who controls and maintains the custom plugin...?

---

I initially started writing the plugin with option (1) in mind, but I'm beginning to lean towards option (2). @awdeorio what do you think? And who should own the plugin code?

**Comment by awdeorio** [comment]
I believe that you're overthinking this!  I use Jekyll for my personal website http://andrewdeorio.com  and I host it on a static file server.

1. Edit a Jekyll template on your local dev machine
2. Execute `jekyll` on the dev machine
3. `rsync` the `_site` folder to your static file server.

That's it!

**Comment by seshrs** [comment]
I can ask them if they have access to Ruby (and I suspect they do). However, I’m not sure that the 482 staff would consider using Primer Spec + Jekyll unless their spec was deployed automatically every time they made a change to it. (With GitHub Pages, GitHub automatically deploys our site when code is committed. Even if not using GitHub pages, at least we are able to configure Travis to perform a deploy step. Unfortunately, I don’t think 482’s current setup makes this easy.)

I think having a plug-in script to provide scaffolding might be a good interim solution. That said, I’ll check with them and post here with updates.

**Comment by awdeorio** [comment]
Here's another solution: a continuous integration script that generates the HTML and rsyncs it to the deployment server.

Here's another solution: an embedded `iframe` that contains the deployed GitHub pages.

There are so many simple solutions here!  It seems like modifying the theme might be overkill.

**Comment by awdeorio** [comment]
After giving this some more thought, maybe it would be better for this theme to be JS-only, and cut out the Jekyll middle man completely.  If that were the case, then *any* website, including one generated using Jekyll could use the theme (and sidebar) by adding a `<script>` tag.

I think the general idea would be to search for header tags, and build a ToC from them.
